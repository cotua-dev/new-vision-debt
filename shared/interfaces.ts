export interface PageProps {
    props: {};
    revalidate: number;
};
