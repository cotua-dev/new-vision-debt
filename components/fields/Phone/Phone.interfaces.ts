import { Dispatch, SetStateAction } from 'react';
import { Step } from '../../Stepper/Stepper.interfaces';

export interface PhoneProps {
    steps: Step[];
    currentStep: number;
    setDisableNextButton: Dispatch<SetStateAction<boolean>>;
    phone: string;
    setPhone: Dispatch<SetStateAction<string>>;
};
