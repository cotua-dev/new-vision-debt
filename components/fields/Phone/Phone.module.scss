@use 'styles/mixins';

div.phone-field-wrapper {
    @include mixins.flexify(center, center, column);

    input.phone-field {
        @include mixins.base-field;
    }

    small.error {
        @include mixins.error;
    }
}
