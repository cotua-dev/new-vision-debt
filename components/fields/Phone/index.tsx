import { ChangeEvent, useState } from 'react';
import NumberFormat from 'react-number-format';
import { PhoneProps } from './Phone.interfaces';
import styles from './Phone.module.scss';

export function Phone(props: PhoneProps): JSX.Element {
    const { steps, currentStep, setDisableNextButton, phone, setPhone } = props;

    // Cite: https://phoneregex.com/
    const usPhoneRegex: RegExp = new RegExp(/1?\W*([2-9][0-8][0-9])\W*([2-9][0-9]{2})\W*([0-9]{4})(\se?x?t?(\d*))?/);

    const [error, setError] = useState('');
    const [showError, setShowError] = useState(false);

    /**
     * Set the value of phone for the step and `phone`
     * @param value String of the phone
     */
    function handlePhoneSetting(value: string): void {
        // Set `phone`
        setPhone(value);

        // Check to see if the value passes the regular expression test
        if (usPhoneRegex.test(value)) {
            // Set and remove error
            setError('');
            setShowError(false);

            // Set the current step's `phone` value
            steps[currentStep].value.phone = value;

            // Make the current step valid
            steps[currentStep].validity = true;

            // Enable the next button
            setDisableNextButton(false);
        } else {
            // Set and show error
            setError('Please provide a phone number');
            setShowError(true);

            // Disable the next button
            setDisableNextButton(true);
        }
    }

    return (
        <div className={styles['phone-field-wrapper']}>
            <NumberFormat
                className={styles['phone-field']}
                mask="_"
                format="+1 ### ### ####"
                type="tel"
                id="phone"
                name="phone"
                value={phone}
                onChange={(event: ChangeEvent<HTMLInputElement>) => handlePhoneSetting(event.target.value)}
            />
            {
                showError &&
                <small className={styles['error']}>{error}</small>
            }
        </div>
    );
}
