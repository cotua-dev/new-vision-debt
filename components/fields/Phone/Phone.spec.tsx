import { render, RenderResult } from '@testing-library/react';
import { Phone } from './index';
import { PhoneProps } from './Phone.interfaces';

describe('Phone Component', (): void => {
    let renderResult: RenderResult;
    const props: PhoneProps = {
        steps: [],
        currentStep: 0,
        setDisableNextButton: () => {},
        phone: '',
        setPhone: () => {},
    };

    beforeAll((): RenderResult => renderResult = render(<Phone {...props}/>));

    test('renders', (): void => {
        const { container } = renderResult;
        expect(container.firstChild).toMatchSnapshot();
    });

    afterAll((): void => renderResult.unmount());
});
