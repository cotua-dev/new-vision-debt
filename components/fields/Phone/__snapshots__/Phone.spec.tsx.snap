// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Phone Component renders 1`] = `
<div
  class="phone-field-wrapper"
>
  <input
    class="phone-field"
    id="phone"
    inputmode="numeric"
    name="phone"
    type="tel"
    value=""
  />
</div>
`;
